Rolling over a data stream alias causes a null pointer exception

When a executing a rollover operation on an alias, if the alias is a data stream alias the operation will fail with a null pointer exception.

Affected versions: Tested on 8.12.0

Reproduction:

```
PUT /_index_template/jimmy-ds-test
{
  "index_patterns": ["jimmy-ds-*"],
  "data_stream": {},
  "template": {
    "aliases": {
      "ds-alias": {
        "is_write_index": true
      }
    }, 
    "settings": {
      "number_of_shards": 1
    },
    "mappings": {
      "properties": {
        "idNumber": {
          "type": "long"
        },
        "@timestamp": {
          "type": "date"
        }
      }
    }
  }
}

PUT /jimmy-ds-test-1/_create/1
{
  "@timestamp": "2024-03-08",
  "idNumber": 4
}

POST /ds-alias/_rollover?error_trace
```

Results in the following exception message:
```
{
  "error": {
    "root_cause": [
      {
        "type": "null_pointer_exception",
        "reason": """Cannot invoke "org.elasticsearch.cluster.metadata.AliasMetadata.writeIndex()" because "aliasMetadata" is null""",
        "stack_trace": """org.elasticsearch.ElasticsearchException$1: Cannot invoke "org.elasticsearch.cluster.metadata.AliasMetadata.writeIndex()" because "aliasMetadata" is null
	at org.elasticsearch.server@8.12.0/org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:673)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.ElasticsearchException.generateFailureXContent(ElasticsearchException.java:601)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.rest.RestResponse.build(RestResponse.java:180)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.rest.RestResponse.<init>(RestResponse.java:140)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.rest.RestResponse.<init>(RestResponse.java:111)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.rest.action.RestActionListener.onFailure(RestActionListener.java:55)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.rest.action.RestCancellableNodeClient$1.onFailure(RestCancellableNodeClient.java:96)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.tasks.TaskManager$1.onFailure(TaskManager.java:214)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:73)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:27)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:39)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:73)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:27)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:73)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:27)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onFailure(ActionListenerImplementations.java:308)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.lambda$doStart$2(TransportMasterNodeAction.java:233)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.acceptException(ActionListenerImplementations.java:186)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onFailure(ActionListenerImplementations.java:191)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.admin.indices.rollover.TransportRolloverAction$RolloverTask.onFailure(TransportRolloverAction.java:251)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService$ExecutionResult.notifyFailure(MasterService.java:975)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService$ExecutionResult.onClusterStateUnchanged(MasterService.java:938)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService.executeAndPublishBatch(MasterService.java:245)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.lambda$run$2(MasterService.java:1626)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListener.run(ActionListener.java:386)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.run(MasterService.java:1623)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService$5.lambda$doRun$0(MasterService.java:1237)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListener.run(ActionListener.java:386)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService$5.doRun(MasterService.java:1216)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NullPointerException: Cannot invoke "org.elasticsearch.cluster.metadata.AliasMetadata.writeIndex()" because "aliasMetadata" is null
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.admin.indices.rollover.MetadataRolloverService.rolloverAlias(MetadataRolloverService.java:202)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.admin.indices.rollover.MetadataRolloverService.rolloverClusterState(MetadataRolloverService.java:117)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.admin.indices.rollover.TransportRolloverAction$RolloverExecutor.executeTask(TransportRolloverAction.java:332)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.admin.indices.rollover.TransportRolloverAction$RolloverExecutor.execute(TransportRolloverAction.java:268)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService.innerExecuteTasks(MasterService.java:1039)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:1004)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService.executeAndPublishBatch(MasterService.java:232)
	... 11 more
"""
      }
    ],
    "type": "null_pointer_exception",
    "reason": """Cannot invoke "org.elasticsearch.cluster.metadata.AliasMetadata.writeIndex()" because "aliasMetadata" is null""",
    "stack_trace": """java.lang.NullPointerException: Cannot invoke "org.elasticsearch.cluster.metadata.AliasMetadata.writeIndex()" because "aliasMetadata" is null
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.admin.indices.rollover.MetadataRolloverService.rolloverAlias(MetadataRolloverService.java:202)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.admin.indices.rollover.MetadataRolloverService.rolloverClusterState(MetadataRolloverService.java:117)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.admin.indices.rollover.TransportRolloverAction$RolloverExecutor.executeTask(TransportRolloverAction.java:332)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.admin.indices.rollover.TransportRolloverAction$RolloverExecutor.execute(TransportRolloverAction.java:268)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService.innerExecuteTasks(MasterService.java:1039)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:1004)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService.executeAndPublishBatch(MasterService.java:232)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.lambda$run$2(MasterService.java:1626)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListener.run(ActionListener.java:386)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.run(MasterService.java:1623)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService$5.lambda$doRun$0(MasterService.java:1237)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListener.run(ActionListener.java:386)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.cluster.service.MasterService$5.doRun(MasterService.java:1216)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
"""
  },
  "status": 500
}
```

--- a/docs/changelog/106708.yaml
+++ b/docs/changelog/106708.yaml
@@ -1,6 +1,6 @@
 pr: 106708
 summary: Improve error message when rolling over DS alias
-area: Data Streams
+area: Data streams
 type: bug
 issues:
  - 106137

--- a/docs/changelog/106708.yaml
+++ b/docs/changelog/106708.yaml
@@ -1,0 +1,6 @@
+pr: 106708
+summary: Improve error message when rolling over DS alias
+area: Data Streams
+type: bug
+issues:
+ - 106137

--- a/server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java
+++ b/server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java
@@ -215,6 +215,16 @@ public class TransportRolloverAction extends TransportMasterNodeAction<RolloverR
             }
         }
 
+        final IndexAbstraction rolloverTargetAbstraction = clusterState.metadata()
+            .getIndicesLookup()
+            .get(rolloverRequest.getRolloverTarget());
+        if (rolloverTargetAbstraction.getType() == IndexAbstraction.Type.ALIAS && rolloverTargetAbstraction.isDataStreamRelated()) {
+            listener.onFailure(
+                new IllegalStateException("Aliases to data streams cannot be rolled over. Please rollover the data stream itself.")
+            );
+            return;
+        }
+
         IndicesStatsRequest statsRequest = new IndicesStatsRequest().indices(rolloverRequest.getRolloverTarget())
             .clear()
             .indicesOptions(IndicesOptions.fromOptions(true, false, true, true))

--- a/server/src/test/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverActionTests.java
+++ b/server/src/test/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverActionTests.java
@@ -552,6 +552,47 @@ public class TransportRolloverActionTests extends ESTestCase {
         }
     }
 
+    public void testRolloverAliasToDataStreamFails() throws Exception {
+        final IndexMetadata backingIndexMetadata = IndexMetadata.builder(".ds-logs-ds-000001")
+            .settings(settings(IndexVersion.current()))
+            .numberOfShards(1)
+            .numberOfReplicas(1)
+            .build();
+        final DataStream dataStream = new DataStream(
+            "logs-ds",
+            List.of(backingIndexMetadata.getIndex()),
+            1,
+            Map.of(),
+            false,
+            false,
+            false,
+            false,
+            IndexMode.STANDARD
+        );
+        Metadata.Builder metadataBuilder = Metadata.builder().put(backingIndexMetadata, false).put(dataStream);
+        metadataBuilder.put("ds-alias", dataStream.getName(), true, null);
+        final ClusterState stateBefore = ClusterState.builder(ClusterName.DEFAULT).metadata(metadataBuilder).build();
+
+        final TransportRolloverAction transportRolloverAction = new TransportRolloverAction(
+            mock(TransportService.class),
+            mockClusterService,
+            mockThreadPool,
+            mockActionFilters,
+            mockIndexNameExpressionResolver,
+            rolloverService,
+            mockClient,
+            mockAllocationService,
+            mockMetadataDataStreamService,
+            dataStreamAutoShardingService
+        );
+
+        final PlainActionFuture<RolloverResponse> future = new PlainActionFuture<>();
+        RolloverRequest rolloverRequest = new RolloverRequest("ds-alias", null);
+        transportRolloverAction.masterOperation(mock(CancellableTask.class), rolloverRequest, stateBefore, future);
+        IllegalStateException illegalStateException = expectThrows(IllegalStateException.class, future::actionGet);
+        assertThat(illegalStateException.getMessage(), containsString("Aliases to data streams cannot be rolled over."));
+    }
+
     private IndicesStatsResponse createIndicesStatResponse(String indexName, long totalDocs, long primariesDocs) {
         final CommonStats primaryStats = mock(CommonStats.class);
         when(primaryStats.getDocs()).thenReturn(new DocsStats(primariesDocs, 0, between(1, 10000)));

